remote-vfs.c

```diff
@@ -0,0 +1,1889 @@
+//
+//                       allow    := do not verify cache-server.  just use it.
+//
+//                       disable  := disable the cache-server and always use
+//                                   the main Git server.
+//
+// Example:
+//
+// $ git -c core.virtualizeobjects=false rev-list --objects --missing=print HEAD~1 >objects.list
+// $ git remote-vfs get-missing <objects.list
+//
+// TODO In this version, we need to turn off "core.virtualizeobjects"
+// TODO when building the list of objects to prevent rev-list from
+// TODO automatically using read-object-hook on them individually
+// TODO (and defeating the whole purpose of this transport).
+//
+// TODO [Currently] we get more credential dialogs than we should.
+//////////////////////////////////////////////////////////////////
+
+#include "cache.h"
+#include "config.h"
+#include "remote.h"
+#include "connect.h"
+#include "strbuf.h"
+#include "walker.h"
+#include "http.h"
+#include "exec-cmd.h"
+#include "run-command.h"
+#include "pkt-line.h"
+#include "string-list.h"
+#include "sideband.h"
+#include "argv-array.h"
+#include "credential.h"
+#include "sha1-array.h"
+#include "send-pack.h"
+#include "protocol.h"
+#include "quote.h"
+#include "transport.h"
+#include "parse-options.h"
+#include "object-store.h"
+#include "json-writer.h"
+#include "tempfile.h"
+#include "oidset.h"
+
+//////////////////////////////////////////////////////////////////
+
+/* Additional error codes beyond HTTP_* in http.h */
+#define RV_ERROR__USAGE                   -1
+#define RV_ERROR__OK                      HTTP_OK
+#define RV_ERROR__UNKNOWN_CACHE_SERVER    20
+#define RV_ERROR__INDEX_PACK_FAILED       21
+#define RV_ERROR__FINALIZE_FAILED         22
+#define RV_ERROR__ERROR                   23 /* unspecified */
+#define RV_ERROR__NOT_FOUND               24
+
+static const char * const main_usage[] = {
+	N_("git remote-vfs [<main_options>] config      [<options>]"),
+	N_("git remote-vfs [<main_options>] get-missing [<options>]"),
+	NULL
+};
+
+static const char *const get_missing_usage[] = {
+	N_("git remote-vfs [<main_options>] get-missing [<options>]"),
+	NULL
+};
+
+//////////////////////////////////////////////////////////////////
+
+enum rv_mode {
+	RV_MODE_AUTO = 0, /* auto-detect */
+	RV_MODE_VFS,      /* force GVFS mode */
+	RV_MODE_SCALAR,   /* force Scalar mode */
+	RV_MODE_NONE,     /* disable product-specific caches */
+};
+
+enum rv_cache_server_use {
+	RV_CSU_VERIFY_DISABLE = 0,  /* verify URL. disable if unknown. */
+	RV_CSU_VERIFY_ERROR,        /* verify URL. error if unknown. */
+	RV_CSU_DISABLE,             /* disable the cache-server, if defined */
+	RV_CSU_ALLOW,               /* allow requests to any cache-server URL */
+};
+
+static struct rv_opts {
+	const char *param__remote_name;
+
+	const char *param__scalar_url; /* Scalar cache-server URL */
+	const char *param__vfs_url;    /* VFS cache-server URL */
+
+	const char *param__scalar_odb_path; /* .scalarCache path */
```

Comment:
I wonder what the value is in having these paths and urls be different based on enlistment type. You source them from different places, yes, but then shouldn't it just be `odb_path` and `url`?

Reply:
I capture all of the possible inputs (in the param__ values) as we startup to try to guess which product we are associated with.  Then a little later I look for a consistent set and pick which product we are actually should be associated with (and set rv_data.* to which ever param__ value we chose).  Perhaps overkill, but it does let us behave differently if necessary.